File: backend/infrahub/graphql/types/permission.py
from __future__ import annotations

from graphene import Field, Int, List, NonNull, ObjectType, String

from infrahub.graphql.types.enums import BranchRelativePermissionDecision


class ObjectPermission(ObjectType):
    kind = Field(String, required=True, description="The kind this permission refers to.")
    view = Field(
        BranchRelativePermissionDecision,
        required=True,
        description="Indicates the permission level for the read action.",
    )
    create = Field(
        BranchRelativePermissionDecision,
        required=True,
        description="Indicates the permission level for the create action.",
    )
    update = Field(
        BranchRelativePermissionDecision,
        required=True,
        description="Indicates the permission level for the update action.",
    )
    delete = Field(
        BranchRelativePermissionDecision,
        required=True,
        description="Indicates the permission level for the delete action.",
    )


class ObjectPermissionNode(ObjectType):
    node = Field(ObjectPermission, required=True)


class PaginatedObjectPermission(ObjectType):
    count = Field(
        Int,
        required=True,
        description="The number of permissions applicable, will be 1 for normal nodes or possibly more for generics",
    )
    edges = Field(List(of_type=NonNull(ObjectPermissionNode), required=True), required=True)

File: frontend/app/tests/fake/permission.ts
import type { Permission } from "@/entities/permission/types";

export type GeneratePermissionOptions = {
  view?: boolean;
  create?: boolean;
  update?: boolean;
  delete?: boolean;
};

export const generatePermission = (options: GeneratePermissionOptions = {}): Permission => {
  const { view = true, create = true, update = true, delete: canDelete = true } = options;

  return {
    view: view ? { isAllowed: true } : { isAllowed: false, message: "View not allowed" },
    create: create ? { isAllowed: true } : { isAllowed: false, message: "Create not allowed" },
    update: update ? { isAllowed: true } : { isAllowed: false, message: "Update not allowed" },
    delete: canDelete ? { isAllowed: true } : { isAllowed: false, message: "Delete not allowed" },
  };
};
